### ISSUE TITLE
Navigating with outlets is confusing: outlets map does not apply to consumed segments

### ISSUE THREAD

**Issue by colltoaction** [angular/angular#13523]
**I'm submitting a ...**
```
[X] bug report
```

**Current behavior**

I have some routes like:

```ts
const routes = [
    { path: "", component: HomeComponent, pathMatch: "full" },
    { path: "profile", component: ProfileComponent },
    { path: "menu", component: MenuComponent, outlet: "sidebar" },
];
```

When the route `http://localhost/(sidebar:menu)` is open, then I'm able to remove the named outlet and go to `http://localhost/` by clicking an anchor like the following:

```html
<a [routerLink]="['', {outlets: {sidebar: null}}]">
    Close
</a>
```

What doesn't work is navigating away from the sidebar **and also** going to another route like  `http://localhost/profile`.

```html
<a [routerLink]="['/profile', {outlets: {sidebar: null}}]">
    Close
</a>
```

The link that is rendered is `http://localhost:3971/profile(sidebar:menu)`

**Expected behavior**

The link that should be rendered is `http://localhost:3971/profile`.

**Repro**

https://github.com/tinchou/router-repro

**Please tell us about your environment:**

Windows 10, Visual Studio 2017

* **Angular version:** 2.3.1 (didn't work with 2.1.0 either)

* **Browser:** all
 
* **Language:** TypeScript 2.0.10

**Comment by tinchou** [comment]
Update: `Router.navigate` doesn't work either. E.g. the following doesn't work:

```ts
this.router.navigate(["/profile", { outlets: { sidebar: null } }]);
```

Workaround: add a click handler and call navigate a second time, **but** only after the first one completes.

```ts
this.router.navigate(["/profile"]).then(() =>
            this.router.navigate([".", { outlets: { sidebar: null } }]));
```

**Comment by vicb** [comment]
Please provide a quick repro. Thanks.

**Comment by tinchou** [comment]
Created a new project with the latest cli and then added the components as seen above. Please take a look: https://github.com/tinchou/router-repro.

**Comment by vicb** [comment]
:+1:

**Comment by WilliamKoza** [comment]
In the vsavkin's book, he says : 

`We can also remove segments by setting them to ‘null’.
router.navigate([{outlets: {popup: null}}])`

But he never set a path and a outlet. In this case, we can use 

<a [routerLink]="[{outlets: {primary: 'profile',sidebar: null}}]">

If it can help

**Comment by tinchou** [comment]
@WilliamKoza it seems to work using the `primary` outlet.  Now I found it in [the docs](https://angular.io/docs/ts/latest/guide/router.html) and it makes sense:

> `outlet`: The name of the `RouterOutlet` used to render the route. For an unnamed outlet, the outlet name is **primary**.

This works perfect for me, thanks a lot (but I believe the issue still stands).

**Comment by nischi** [comment]
same issue here. i try to close the a modal window. I have a outlet called "modal" and there i load my modal window component. If i want to close it (without knowing where i'm) with modal: null it routes to the modal again.

**Comment by ragv1** [comment]
same issue. i have a named outlet call popup, just like in the docs. when i try to close it with:
`this.router.navigate([{ outlets: {popup: null }}]);`
It just does nothing.

**Comment by Almar** [comment]
@ragv1, have you tried ```this.router.navigate(['', { outlets: {popup: null }}]);```?

**Comment by zhumeisongsong** [comment]
`this.router.navigate([{outlets: {popup: null}}], {relativeTo: this.route}); `
Import ActivatedRoute fixed my issue.

**Comment by bennadel** [comment]
I am trying to get a modal window (ish) to close itself and it doesn't seem to do anything. I've tried using the `""` as the first part of the `.navigate()`, but no luck. And, I have `enableTracing` turned on, and it doesn't show any activity. It's like the `.navigate()` action isn't even doing anything. 

For reference:

```js
this.router.navigate(
	[
		"",
		{
			outlets: {
				inbox: null // <---- Doesn't seem to do anything from WITHIN `inbox` route.
			}
		}
	],
	{
		relativeTo: this.activatedRoute
	}
);
```

**Comment by bennadel** [comment]
Also, if I try to use `"../"` as the first portion - `this.router.navigate( "../" ... )`, then the secondary outlet *does close*; however, it leaves it in the URL as `//inbox:)`. And, at that point, if I refresh the browser page, I get an error that the primary route can no longer match.

**Comment by bennadel** [comment]
For what it's worth, I was able to get this working by using the **parent component's** `ActivatedRoute` instance (`this.activatedRoute.parent`) in the `router.navigate()` method. I [wrote up my approach here](https://www.bennadel.com/blog/3351-closing-secondary-router-outlet-views-from-within-the-named-route-view-components-in-angular-4-4-4.htm).

**Comment by challe** [comment]
@bennadel My saviour

**Comment by bennadel** [comment]
@challe my pleasure :) It's a _partial solution_. It still basically means the `[routerLinker]` can't change a secondary outlet that it doesn't own; which is unfortunate because it means that you can't have natural HREF values (and that you have to create special methods in those cases). I think this might end up being a significant blocker for me :|

**Comment by picheli20** [comment]
I've opened a similar issue (https://github.com/angular/angular/issues/19547), didn't saw this one before ;p

I did some work around similar to the @tinchou but always going from root, it could be improved using promise instead of `setTimeout` for sure, but basically I do each step at time.

```js
  justGo (path) {
    this.router.navigate(['/']);
    [].map.call(arguments, (arg, index) => {
      setTimeout(() => this.router.navigate(arg), 10 * (index + 1));
    });
  }
```

And I use like that:

```html
(click)="this.localeService.justGo(['/promotions'], ['', { outlets: { modal: [...] } }])"
```

**Comment by radomeit** [comment]
I wanted to change the route and at the same time change an aux route as described above with the only difference that i wanted to show a dialog on route change instead of hiding it.
It finally went with this workaround:
`        this.router.navigate([
            {
                outlets: {
                    'primary': ['home'],
                    'dialogs': ['mydialog', dialogparams]
                }
            }
        ]);
`

The key was to not set '/home' as the navigation target but instead move it to the outlets section as 'primary' outlet.

**Comment by Hypenate** [comment]
Came across the same issue I guess...
I'm trying to generate this URL: ./Home/(FooOutlet:Foo/(BarOutlet:Bar)).

`this.router.navigate(['Home', {outlets: {primary: '', FooOutlet 'Foo', BarOutlet: 'Bar'}}]);`
equals to /Home/(//FooOutlet:Foo:Foo//BarOutlet:Bar)

`this.router.navigate([{ outlets: { 'primary': ['Home'], 'FooOutlet ': ['Foo'], 'BarOutlet': ['Bar'] } }]);`
equals to /Home(FooOutlet:Foo//BarOutlet:Bar) and an error that it cannot find Bar

`this._router.navigate(['Home', { outlets: { home: ['Requests'] } }]).then(() =>
   this._router.navigate([{ outlets: { request: ['EmergencyRequest'] } }]));`
equals to Home/(FooOutlet:Foo)(BarOutlet:Bar) and an error that it cannot find Bar

Is it me missing something or is Angular unable to construct a URL that goes 2 levels/router-outlets deep?

**Comment by atscott** [comment]
Confirmed with repro in v9: https://stackblitz.com/edit/angular-ivy-1xrrur?file=src%2Fapp%2Fapp.module.ts

**Comment by atscott** [comment]
I've been working on the outlet code a bit recently and have finally come to a relatively decent understanding of how this all works. I've also closed a bunch of issues as duplicates of this one. It's an _extremely_ common issue and there are two main problems:
1. The documentation doesn't set clear expectations. In most cases, this is actually working as intended, even if it might not be very ergonomic. 
2. There isn't currently an easy way to get the behavior that many people often want: Apply the outlet command to any and all existing matches in the url tree.

Here's a full explanation / example of what's happening (you can also see a similar explanation in another report [here](https://github.com/angular/angular/issues/15718#issuecomment-717434043)):
```
const routes = [
  {
    path: "taxonomy",
    component: Labels,
    children: [
      {
        path: ":discriminantId/create_label/:type",
        component: LabelsCreate,
        outlet: "create"
      },
      { outlet: "browse", path: "browse", component: LabelsBrowse }
    ]
  }
];
```
Let's say we're in the state `taxonomy/(browse:browse)` and trying to navigate to `/taxonomy/(create:type/discriminantId/create_label/getChildLabelNodeType//browse:browse)`

You have to think of the routing configuration (and parsed url) as a tree and the navigation commands as targeting specific "levels" in that tree.

So if you were to do
```
router.navigate([{
  outlets: {
    create: [
      type, discriminantId, 'create_label', this.getChildLabelNodeType(type)
    ],
    browse: 'browse'
  },
}]
```
These commands are all targeting the root level of the tree and the generated url applies them all to the root level, giving `/(create:type/discriminantId/create_label/getChildLabelNodeType//browse:browse)'`

All `routerLink` routes are generated with `relativeTo: this.route`, so navigating to an "adjacent" outlet is a little difficult from here because the `relativeTo` already consumes the segment you want to target.

If you're navigating _from_ a component within the `browse` route, you'll want to use `ActivatedRoute#parent`, otherwise the tree creation will consume both level 0 (taxonomy) and level 1 (browse:browse), then try to use your commands to apply the route to the segments at level 2.

Example: https://stackblitz.com/edit/angular-ivy-jxeecz?file=src%2Fapp%2Faside.component.ts

Long story short, you have to be aware of what segments of the route are being consumed before your commands are interpreted.

**Comment by atscott** [comment]
Closed by #40272
